### DISCUSSION TITLE
Vector Data recognition

### DISCUSSION THREAD

**Discussion by slgaficon** [ImageMagick/ImageMagick#2660] [Help]
We have an algorithm to detect if an EPS file contains vector data only.
This starts with calling:
`convert my_file.eps profile.8bim`

IM 7 has another behavior than IM 6 which we have been using up to now.
- IM 6: create zero byte profile what indicates that vector data is present.
- IM 7: create 242 bytes profile

So, IM 7 breaks our current analysis because is always writes data to the profile on disk. 
Are there any ideas how in IM 7 using the 8bim profile a recognition of pure vector data can be done?

**Comment by slgaficon** [comment]
Example content of a pure vector EPS file 8BIM listed by identify:

```
Image: C:\Users\sl\Desktop\profile-1.8bim
  Format: 8BIM (Photoshop resource format)
  Class: DirectClass
  Geometry: 1x1+0+0
  Units: Undefined
  Colorspace: sRGB
  Type: Bilevel
  Base type: Undefined
  Endianess: Undefined
  Depth: 16/1-bit
  Channel depth:
    Red: 1-bit
    Green: 1-bit
    Blue: 1-bit
  Channel statistics:
    Pixels: 1
    Red:
      min: 65535  (1)
      max: 65535 (1)
      mean: 65535 (1)
      standard deviation: 0 (0)
      kurtosis: -3
      skewness: 0
      entropy: 0
    Green:
      min: 65535  (1)
      max: 65535 (1)
      mean: 65535 (1)
      standard deviation: 0 (0)
      kurtosis: -3
      skewness: 0
      entropy: 0
    Blue:
      min: 65535  (1)
      max: 65535 (1)
      mean: 65535 (1)
      standard deviation: 0 (0)
      kurtosis: -3
      skewness: 0
      entropy: 0
  Image statistics:
    Overall:
      min: 65535  (1)
      max: 65535 (1)
      mean: 65535 (1)
      standard deviation: 0 (0)
      kurtosis: -3
      skewness: 0
      entropy: 0
  Colors: 1
  Histogram:
         1: (65535,65535,65535) #FFFFFFFFFFFF white
  Rendering intent: Perceptual
  Gamma: 0.454545
  Chromaticity:
    red primary: (0.64,0.33)
    green primary: (0.3,0.6)
    blue primary: (0.15,0.06)
    white point: (0.3127,0.329)
  Matte color: grey74
  Background color: white
  Border color: srgb(223,223,223)
  Transparent color: none
  Interlace: None
  Intensity: Undefined
  Compose: Over
  Page geometry: 1x1+0+0
  Dispose: Undefined
  Iterations: 0
  Compression: Undefined
  Orientation: Undefined
  Properties:
    date:create: 2020-09-30T18:51:34+02:00
    date:modify: 2020-09-30T18:51:34+02:00
    signature: cc143326a2646c605ea66139d7b440df7cbde18c050f1f8cf4dd30f42cfe7123
  Clipping path: 
<?xml version="1.0" encoding="iso-8859-1"?>
<svg xmlns="http://www.w3.org/2000/svg" width="1" height="1">
<g>
<path fill-rule="evenodd" style="fill:#000000;stroke:#000000;stroke-width:0;stroke-antialiasing:false" d="
M 1 0
C 1 0.285928 0.997151 0.566514 0.997151 0.852564
C 0.997151 0.901709 0.997151 0.950855 0.997151 1
L 0 1
C 0 0.715812 0 0.431624 0 0.147436
C 0 0.0982906 0 0.0491453 0 0
L 1 0 Z
"/>
</g>
</svg>

  Profiles:
    Profile-8bim: 242 bytes
  Artifacts:
    verbose: true
  Tainted: False
  Filesize: 242B
  Number pixels: 1
  Pixels per second: 1661P
  User time: 0.000u
  Elapsed time: 0:01.000
  Version: ImageMagick 7.0.8-27 Q16 x64 2019-02-09 http://www.imagemagick.org
```

An idea would be to look at "Number Pixels" or "Geometry". 1 Pixel seems to indicate that no raster image data is present.
Can anyone confirm?

**Comment by snibgo** [comment]
I think you are asking how to detect whether an EPS file contains any embedded raster (bitmap) images.

I don't see why the presence or absence of 8BIM data would indicate that. The size of the image at any particular density is also not a guide.

IM doesn't know (or care) if it contains embedded raster images, because it delegates the rasterization to Ghostscript. So the answer would come from GS. Perhaps it has some debugging or status output that says whether any rasters were present.

**Reply by slgaficon** [reply]
You are right. It is about detecting if a EPS contains raster images.

In IM 6 it was reliable for rater image detection to extract and check for an 8BIM profile. The output shown in my previous post comes from identiy applied on the .8BIM file.
So these 242 bytes only show what Adobe writes as additional metadata for EPS files. My idea was - after this file is written always with IM 7 - to analyze data an conclude if a full vector image is present.

**Comment by t246246** [comment]
If you are using unix-like environment, file command might be your help?
(See https://github.com/ImageMagick/ImageMagick/issues/2666#issuecomment-703038650 for example.)
Standard eps file contains only PostScript part, whereas Adobe one probably contains tiff.

If there is no tiff part, then extract PostScript part and check that file contains image somehow, for example convert it to pdf and use pdfimages.  (But as far as I experiened, some pdf file may contains image that pdfimages cannot detect. https://bugs.freedesktop.org/show_bug.cgi?id=91734)
